pyof/foundation/basic_types.py

```diff
@@ -421,10 +421,23 @@ def __init__(self, value=b''):  # noqa
         Raises:
             ValueError: If given value is not bytes.
         """
-        if not isinstance(value, bytes):
-            raise ValueError('BinaryData must contain bytes.')
+        value = self._pack_if_necessary(value)
         super().__init__(value)
 
+    @staticmethod
+    def _pack_if_necessary(value):
+        if hasattr(value, 'pack') and callable(value.pack):
+            value = value.pack()
+        elif not value:
+            value = b''
+
+        if not isinstance(value, bytes):
+            msg = 'BinaryData value must contain bytes or have pack method; '
+            msg += 'Received type {} value: {}'.format(type(value), value)
```

Comment:
This is a better message for debugging. It is a good practice to put the value between double quotes so the user can visualize an empty value. When changing this, you can also fix a minor issue: use period or lowercase "Received".